File: src/utils/riskScoring.js
// Mirrors backend/risk_scoring.py — keep both in sync if the scale changes.

export const LEVELS = ['VL', 'L', 'M', 'H', 'VH']
export const LEVEL_LABEL = { VL: 'Very low', L: 'Low', M: 'Medium', H: 'High', VH: 'Very high' }
export const SCALE_UP = { VL: 1, L: 2, M: 3, H: 4, VH: 5 }
// Cost is inverted on purpose: a cheap mechanism (VL cost) scores high, an
// expensive one (VH cost) scores low — cost is the one axis where "less"
// is the good outcome.
export const SCALE_COST = { VL: 5, L: 4, M: 3, H: 2, VH: 1 }

// Placeholder tiering, pending a proper management-scorecard method — net
// ranges roughly -4..+4. Only risks at or above this threshold count as
// "well managed" enough to name in an investor-facing Generate Plan.
export const WELL_MANAGED_THRESHOLD = 1

export function exposureScore(risk) {
  return (SCALE_UP[risk.probability] + SCALE_UP[risk.impact]) / 2
}

export function mechanismScore(mechanism) {
  return (SCALE_UP[mechanism.capacity] + SCALE_COST[mechanism.cost]) / 2
}

export function mitigationScore(mechanisms) {
  if (!mechanisms || mechanisms.length === 0) return 0
  return mechanisms.reduce((sum, m) => sum + mechanismScore(m), 0) / mechanisms.length
}

export function netScore(risk, mechanisms) {
  if (!mechanisms || mechanisms.length === 0) return 0
  return mitigationScore(mechanisms) - exposureScore(risk)
}

export function managementTier(net, hasMechanisms) {
  if (!hasMechanisms) return 'unmanaged'
  if (net >= WELL_MANAGED_THRESHOLD) return 'well_managed'
  if (net >= 0) return 'partially_managed'
  return 'unmanaged'
}

export const TIER_LABEL = {
  unmanaged: 'Unmanaged',
  partially_managed: 'Partially managed',
  well_managed: 'Well managed',
}

export const TIER_COLOR = {
  unmanaged: '#FCA5A5',
  partially_managed: '#F2C77D',
  well_managed: '#9FE7C6',
}

File: backend/risk_scoring.py
"""Risk Analysis scoring — the same 5-point scale and formulas the frontend
uses (see riskScoring.js), duplicated here only because the Generate Plan
endpoint needs to decide server-side which risks qualify as "well managed"
before it will write about them. Keep both in sync if the scale changes.
"""

SCALE_UP = {'VL': 1, 'L': 2, 'M': 3, 'H': 4, 'VH': 5}
# Cost is inverted on purpose: a cheap mechanism (VL cost) scores high, an
# expensive one (VH cost) scores low — cost is the one axis where "less" is
# the good outcome.
SCALE_COST = {'VL': 5, 'L': 4, 'M': 3, 'H': 2, 'VH': 1}

# Placeholder tiering, pending the user's own management-scorecard method —
# net = mitigation - exposure, ranges roughly -4..+4. Only risks at or above
# this threshold are considered "well managed" enough to name in an
# investor-facing plan.
WELL_MANAGED_THRESHOLD = 1.0


def exposure_score(risk) -> float:
    return (SCALE_UP.get(risk.probability, 3) + SCALE_UP.get(risk.impact, 3)) / 2


def mechanism_score(mechanism) -> float:
    return (SCALE_UP.get(mechanism.capacity, 3) + SCALE_COST.get(mechanism.cost, 3)) / 2


def mitigation_score(mechanisms) -> float:
    if not mechanisms:
        return 0.0
    return sum(mechanism_score(m) for m in mechanisms) / len(mechanisms)


def net_score(risk, mechanisms) -> float:
    if not mechanisms:
        return 0.0
    return mitigation_score(mechanisms) - exposure_score(risk)


def management_tier(net: float, has_mechanisms: bool) -> str:
    if not has_mechanisms:
        return 'unmanaged'
    if net >= WELL_MANAGED_THRESHOLD:
        return 'well_managed'
    if net >= 0:
        return 'partially_managed'
    return 'unmanaged'


def is_well_managed(risk, mechanisms) -> bool:
    return management_tier(net_score(risk, mechanisms), bool(mechanisms)) == 'well_managed'
